src/asmcnc/job/yetipilot/yetipilot.py

```diff
@@ -1,59 +1,314 @@
+
+            if not self.using_advanced_profile:
+                if abs(int(self.m.s.spindle_speed) - self.jd.grbl_mode_tracker[0][2]) > 500:
+                    self.adjust_spindle_speed(self.jd.grbl_mode_tracker[0][2])
+
+    # SPINDLE SPEED ADJUSTMENTS
+    def adjust_spindle_speed(self, current_rpm):
+        total_override_required = (self.target_spindle_speed / current_rpm) * 100
+        current_override = self.m.s.speed_override_percentage
+        difference = total_override_required - current_override
+        adjustments = get_adjustment(difference)
+        self.do_spindle_adjustment(adjustments)
+
+    def do_spindle_adjustment(self, adjustments):
+        for i, adjustment in enumerate(adjustments):
+            if self.m.s.speed_override_percentage == 200 and adjustment > 0:
+                break
+
+            if self.m.s.speed_override_percentage + adjustment > 200:
+                adjustment = 1
+
+            if adjustment == 10:
+                Clock.schedule_once(lambda dt: self.feed_override_wrapper(self.m.speed_override_up_10),
+                                    i * self.override_command_delay)
+            elif adjustment == 1:
+                Clock.schedule_once(lambda dt: self.feed_override_wrapper(self.m.speed_override_up_1),
+                                    i * self.override_command_delay)
+            elif adjustment == -10:
+                Clock.schedule_once(lambda dt: self.feed_override_wrapper(self.m.speed_override_down_10),
+                                    i * self.override_command_delay)
+            elif adjustment == -1:
+                Clock.schedule_once(lambda dt: self.feed_override_wrapper(self.m.speed_override_down_1),
+                                    i * self.override_command_delay)
+
+    def stop_and_show_error(self):
+        self.disable()
+        self.m.stop_for_a_stream_pause('yetipilot_low_feed')
+
+    def check_if_feed_too_low(self):
+        if self.m.s.feed_override_percentage == 10:
+            self.stop_and_show_error()
+        self.waiting_for_feed_too_low_decision = False
+
+    def feed_override_wrapper(self, feed_override_func):
+        if self.use_yp and self.m.s.is_job_streaming and \
+                not self.m.is_machine_paused and "Alarm" not in self.m.state():
+            feed_override_func()
+
+    def do_adjustment(self, adjustments):
+        for i, adjustment in enumerate(adjustments):
+            if i == self.override_commands_per_adjustment:
+                break
+
+            if self.m.s.feed_override_percentage == 200 and adjustment > 0:
+                break
+
+            if self.m.s.feed_override_percentage + adjustment > 200 and adjustment == 10:
+                adjustment = 1
+
+            if self.m.s.feed_override_percentage + adjustment < 10:
+                if not self.waiting_for_feed_too_low_decision:
+                    Clock.schedule_once(lambda dt: self.check_if_feed_too_low(), 4)
```

Comment:
This unchecked clock has potential to throw up a screen at a semi-random time, i.e. if a user has already stopped or paused the job or the spindle has overloaded, then the job & YP could have been disabled but this screen can still be thrown up. Needs some kind of checks that only throw the error if the machine/console is in the right state.